src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,292 @@
+
+                    byte type, code;
+                    unsafe
+                    {
+                        fixed (byte* bytesPtr = receiveBuffer)
+                        {
+                            int icmpHeaderOffset = ipHeaderLength;
+                            IcmpHeader receivedHeader = *((IcmpHeader*)(bytesPtr + icmpHeaderOffset)); // Skip IP header.
+                            type = receivedHeader.Type;
+                            code = receivedHeader.Code;
+
+                            if (identifier != receivedHeader.Identifier
+                                || type == (byte)IcmpV4MessageType.EchoRequest
+                                || type == (byte)IcmpV6MessageType.EchoRequest) // Echo Request, ignore
+                            {
+                                continue;
+                            }
+                        }
+                    }
+
+                    sw.Stop();
+                    long roundTripTime = sw.ElapsedMilliseconds;
+                    int dataOffset = ipHeaderLength + IcmpHeaderLengthInBytes;
+                    // We want to return a buffer with the actual data we sent out, not including the header data.
+                    byte[] dataBuffer = new byte[bytesReceived - dataOffset];
+                    Array.Copy(receiveBuffer, dataOffset, dataBuffer, 0, dataBuffer.Length);
+
+                    IPStatus status = isIpv4
+                                        ? IcmpV4MessageConstants.MapV4TypeToIPStatus(type, code)
+                                        : IcmpV6MessageConstants.MapV6TypeToIPStatus(type, code);
+
+                    return new PingReply(address, options, status, roundTripTime, dataBuffer);
+                }
+
+                // We have exceeded our timeout duration, and no reply has been received.
+                sw.Stop();
+                throw new PingException(SR.net_ping_reply_timeout);
+            }
+        }
+
+        private async Task<PingReply> SendWithPingUtility(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        {
+            bool isIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            string pingExecutable = isIpv4 ? s_ipv4PingFile : s_ipv6PingFile;
+
+            StringBuilder sb = new StringBuilder();
+            sb.Append("-c 1"); // Just send a single ping ("count = 1")
+
+            // The command-line flags for "Do-not-fragment" and "TTL" are not standard.
+            // In fact, they are different even between ping and ping6 on the same machine.
+
+            // The ping utility is not flexible enough to specify an exact payload.
+            // But we can at least send the right number of bytes.
+            sb.Append(" -s ");
+            sb.Append(buffer.Length);
+
+            sb.Append(' ');
+            sb.Append(address.ToString());
+
+            string processArgs = sb.ToString();
+            ProcessStartInfo psi = new ProcessStartInfo(pingExecutable, processArgs);
```

Comment:
Security: You're executing a command from the path.  If the user context happens to have put any custom directories ahead of the system directories in their path AND they happen to be writable by this process then there's the potential for an arbitrary code execution.  If it's hosted in ASP.Net, then now it becomes remote code execution.

We'll need to have an explicit list of paths to try, either built from the PAL (to be only one thing) or sequenced.

`/bin/ping` feels like it should always be the right answer.  If that happens to not be true on OSX/FreeBSD/NetBSD/etc then we can do other fallbacks..